src/uu/ln/src/ln.rs

```diff
@@ -277,8 +284,36 @@ fn link_files_in_dir(files: &[PathBuf], target_dir: &PathBuf, settings: &Setting
     }
 }
 
+fn relative_path(src: &PathBuf, dst: &PathBuf) -> PathBuf {
+    let abssrc = canonicalize(src, CanonicalizeMode::Normal).unwrap();
+    let absdst = canonicalize(dst, CanonicalizeMode::Normal).unwrap();
+    let suffix_pos = abssrc
+        .components()
+        .zip(absdst.components())
+        .take_while(|(s, d)| s == d)
+        .count();
+
+    let mut result = PathBuf::new();
+    absdst
+        .components()
+        .skip(suffix_pos + 1)
+        .map(|_| result.push(".."))
+        .last();
+    abssrc
+        .components()
+        .skip(suffix_pos)
+        .map(|x| result.push(x.as_os_str()))
+        .last();
+    result
+}
+
 fn link(src: &PathBuf, dst: &PathBuf, settings: &Settings) -> Result<()> {
     let mut backup_path = None;
+    let source = if settings.relative {
+        relative_path(&src, dst)
+    } else {
+        src.clone()
```

Comment:
You can probably avoid this clone by making `source` a `Cow<‘_, Path>` and editing a few functions to take `&Path` instead of `&PathBuf`.